examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_hir;
+extern crate rustc_middle;
+extern crate rustc_span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_hir::{Body, FnDecl, intravisit::FnKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_span::{
+    Span,
+    def_id::{DefId, LocalDefId},
+};
+
+dylint_linting::declare_late_lint! {
+    pub ARG_ITER,
+    Warn,
+    "functions taking `Iterator` trait bounds when `IntoIterator` would be more flexible"
+}
+
+impl<'tcx> LateLintPass<'tcx> for ArgIter {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _body: &'tcx Body<'tcx>,
+        span: Span,
+        id: LocalDefId,
+    ) {
+        // Don't lint functions generated by macros (e.g. derive, attribute macros)
+        if span.from_expansion() {
+            return;
+        }
+
+        let Some(iterator_def_id) = cx.tcx.lang_items().iterator_trait() else {
+            return;
+        };
+        let Some(sized_def_id) = cx.tcx.lang_items().sized_trait() else {
+            return;
+        };
+
+        let generics = cx.tcx.generics_of(id);
+        let predicates = cx.tcx.predicates_of(id);
+
+        if generics.own_params.is_empty() {
+            return;
+        }
+
+        let fn_sig = cx.tcx.fn_sig(id).skip_binder();
+
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            let input_ty_skipped_binder: &ty::Ty = input_ty.skip_binder(); // Skip binder once
+            if let ty::TyKind::Param(param_ty) = input_ty_skipped_binder.kind() {
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter is bound only by Iterator
+                if is_param_bound_only_by_iterator(
+                    cx,
+                    predicates,
+                    *input_ty_skipped_binder,
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) && !is_param_used_in_other_trait_args(
+                    cx,
+                    predicates,
+                    *input_ty_skipped_binder,
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) {
+                    let input_hir_ty = &decl.inputs[i];
+                    span_lint_and_help(
+                        cx,
+                        ARG_ITER,
+                        input_hir_ty.span,
+                        "parameter type has `Iterator` bound",
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
```

Comment:
```suggestion
                            param_ty.name
```
I think you can get the parameter [name](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/struct.ParamTy.html#structfield.name) from the `ParamTy`, making the use of generics unnecessary.